processing/src/main/java/io/druid/segment/data/GenericIndexedWriter.java

```diff
@@ -107,35 +177,130 @@ public void close() throws IOException
         headerOut.getCount()
     );
     Preconditions.checkState(
-        numBytesWritten < Integer.MAX_VALUE, "Wrote[%s] bytes, which is too many.", numBytesWritten
+        numBytesWritten < fileSizeLimit, "Wrote[%s] bytes, which is too many.", numBytesWritten
     );
 
-    OutputStream metaOut = ioPeon.makeOutputStream(makeFilename("meta"));
-
-    try {
+    try (OutputStream metaOut = ioPeon.makeOutputStream(makeFilename("meta"))) {
       metaOut.write(0x1);
       metaOut.write(objectsSorted ? 0x1 : 0x0);
-      metaOut.write(Ints.toByteArray((int) numBytesWritten + 4));
-      metaOut.write(Ints.toByteArray(numWritten));
+      metaOut.write(Ints.toByteArray(Ints.saturatedCast(numBytesWritten + 4)));
+      metaOut.write(Ints.toByteArray(Ints.saturatedCast(numWritten)));
     }
-    finally {
-      metaOut.close();
+  }
+
+  private void closeMultiFiles() throws IOException
+  {
+    headerOutLong.close();
+    Preconditions.checkState(
+        headerOutLong.getCount() == (numWritten * 8),
+        "numWritten[%s] number of rows should have [%s] bytes written to headerOutLong, had[%s]",
+        numWritten,
+        numWritten * 8,
+        headerOutLong.getCount()
+    );
+    Preconditions.checkState(
+        headerOutLong.getCount() < Integer.MAX_VALUE,
```

Comment:
Whatever. Integer.MAX_VALUE itself is not a valid array size, so heap byteBuffer couldn't be of this size. Generally we shouldn't copy direct BBs from `GenericIndexed` to heap, but not sure. Also authors of 3rd party extensions could do that.